### ISSUE TITLE
sd_pid_notify_with_fds doesn't work on shm or memfd even though the API returns OK

### ISSUE THREAD

**Issue by utezduyar** [systemd/systemd#5606]
### Submission type

  - [x] Bug report

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> 232

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

> Debian

### In case of bug report: Expected behaviour you didn't see

> We are sending fd to systemd with sd_pid_notify_with_fds. The API returns success but later on in the logs I see that systemd is complaining about the received file descriptor. The problem is not there when I am sending socket fd. Code seems to be failing on epoll_ctl.

### In case of bug report: Unexpected behaviour you saw

`systemd[1]: state.service: Failed to add fd to store: Operation not permitted`

### In case of bug report: Steps to reproduce the problem

```
#include <sys/mman.h>
#include <sys/stat.h>      
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <systemd/sd-daemon.h>
#include <string.h>

#define len 10

void error(char *s){
  perror(s);
  exit(errno);
}

int main() {
  int fd;
  fd = shm_open("/tempsharedmem", O_RDWR | O_CREAT, S_IRWXU | S_IRWXG | S_IRWXO); 
  if (fd < 0) error("Failed to open memory object");
  if (ftruncate(fd, len) < 0)	error("Failed to allocate shared memory");
  sd_pid_notify(0, 0, "READY=1");
  if (sd_pid_notify_with_fds(0, 0, "FDSTORE=1\n", &fd, 1) < 1) error("Failed to pass the fd to PID 1");
  if (shm_unlink("/tempsharedmem") == -1) error("Failed to unlink fd");
  printf("Exited properly");
  return 0;
}

```

**Comment by poettering** [comment]
hmm, indeed...

We watch the fds for disconnects. I figure we should have a smarter fallback for fds which know no concept of disconnect, and simply not do this for them.

Fixing this should be pretty straightforward.

**Comment by keszybz** [comment]
@poettering Can you remind me why we require fds to be stored to be pollable? I know this was discussed, I even added it to docs in 99bdcdc7fc32def6cd6eab69c3869661b4b1888b, but I'm now not sure if this requirement was just because it was implemented this way or if there was a more fundamental reason.

**Comment by poettering** [comment]
@keszybz we automatically remove fds from the store as soon as we get a POLLHUP on it, so that we don't keep fds open referring to dead objects forever. 

But I figure we should relax the requirement, I figure it's OK to permit fd objects in them that'll never get POLLHUP anyway...

**Comment by poettering** [comment]
fix waiting in #6160
